src/compiler/checker.ts

```diff
@@ -4627,7 +4628,7 @@ namespace ts {
             // function/class/{} assignments are fresh declarations, not property assignments, so only add prototype assignments
             const specialDeclaration = getAssignedJavascriptInitializer(symbol.valueDeclaration);
             if (specialDeclaration) {
-                return getWidenedLiteralType(checkExpressionCached(specialDeclaration));
+                return getWidenedLiteralType(checkExpressionCached(specialDeclaration, CheckMode.NoDeferredCheck));

```

Comment:
So. `CheckMode.NoDeferredCheck` isn't falsy. Which means this call to `checkExpressionCached` is really a roundabout call to `checkExpression`.

Reply:
Yes, nice catch. Fixed.

Aside: This means that, originally, the second check of the method body must have gone through checkExpression already or it would have gotten the resolved type and avoided it. This leads to an invariant: once you call checkExpressionCached you should only ever call checkExpressionCached afterward, not checkExpression. Or checkExpression should check links.resolvedType before doing any work.

What's more, it's not *in general* safe to check any node twice because of side effects to deferredNodes. Maybe expensive asserts in checkDeferredNodes would actually be worth it...

Later: I added the resolvedType check in checkExpression and the only change it makes to tests is the order that a couple of unions print in. So it might be a worthwhile change &mdash; even better than the current one, in fact.